File: leetcode/typescript/000094.binary-tree-inorder-traversal.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val
		this.left = left === undefined ? null : left
		this.right = right === undefined ? null : right
	}
}

// Recursive
function inorderTraversal(root: TreeNode | null): number[] {
	const res: number[] = []

	function walk(node: TreeNode | null) {
		if (node) {
			walk(node.left)
			res.push(node.val)
			walk(node.right)
		}
	}

	walk(root)

	return res
}
// Time complexity: O(n)
// Space complexity: O(n)

// Imperative using stack
function inorderTraversalWithStack(root: TreeNode | null): number[] {
	const res: number[] = []
	const stack: TreeNode[] = []
	let curr = root

	while (curr || !(stack.length === 0)) {
		while (curr) {
			stack.push(curr)
			curr = curr.left
		}

		curr = stack.pop()!
		res.push(curr.val)
		curr = curr.right
	}

	return res
}
// Time complexity: O(n)
// Space complexity: O(n)

File: leetcode/python/000094.binary-tree-inorder-traversal.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        values = []

        def dfs(node):
            if node is None:
                return

            dfs(node.left)
            values.append(node.val)
            dfs(node.right)

        dfs(root)
        return values

    def inorderTraversal2(self, root: Optional[TreeNode]) -> List[int]:
        values = []
        stack = []
        cur = root

        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left

            cur = stack.pop()
            values.append(cur.val)

            cur = cur.right

        return values
